### DISCUSSION TITLE
Change definition of "home" location from "take-off" to "radio" location for radios with built in GPS receivers.

### DISCUSSION THREAD

**Discussion by dzaro-dev** [iNavFlight/inav#8575] [General]
## Current Behavior

At the moment, the HOME location used by INAV for each flight is defined as the location where the aircraft takes off. 

## Desired Behavior

The user should have the option between two RTH modes:

1. `RTH` (return to home), where "home" is defined as the takeoff location of the drone / plane
2. `RTO` (return to operator). where the "operator" or defined by the GPS coordinates of the transmitter used by the operator

Radios like the popular Radiomaster TX16S allow adding a GPS receiver to one of the two available USART ports using cheap GPS recivers like the Beitian BN-220. My suggestion is to have two additional telemetry items called "`GPSt`" (GPS of Transmitter) and "`ALTt`" (Altitude of Transmitter) which would report the GPS location and the altitide of the transmitter back to the drone / plane. 

This would allow for the addition of two or more functional navigation modes such as:

1. `RTO`, i.e. return to operator, as described above.
2. `Follow-me with fixed altitude`: The drone / plane returns to the operator's location but does not land. Drones would then hover at that location, whilst planes would step into a temprary loiter mode until the location of the operator changes by either the specified loiter radius, or an independently defined variable. The height would be defined as a distance above the operator, the calculated RTH altitude after take-off or an independently defined variable. 
3. `Follow-me with adaptive altitude`: This would be similar to the simple "follow-me" mode, with the crucial difference that an altitude offset between the drone / plane and the transmitter would be defined by an offset value, rather than a static value.

## Suggested Solution

The solution is to liaise with the EdgeTX and ELRS development teams, to add Trasmitter Telemetry for bi-directional telemetry between to accomodate for those two additional values, i.e. "GPSt" and "ALTt", to be transmitted to the drone / plane.

## Who does this impact? Who is this for?

I think that this would be a great feature to have for anyone who is looking to fly autonomously whilst being on the move. This "follow-me" fature is used a lot by DJI pilots who fly drones with vision based "follow-me". It would open up a whole new set of possibilities for INAV and the usega of it and it would make a great feature that other RTF drones have.

## Additional context
`RTH` and `FAILSAFE` behaviour would remain the same as now and no changes are required to those two modes.
Something similar has already been implemented by @rotorman (Risto) for the Yaapu Telemetry script [here](https://www.rcgroups.com/forums/showthread.php?3840599-Blog-6-Modified-Yaapu-telemetry-widget-to-support-MavSDK-native-MAVLink-on-OpenTX), but it only happens on the transmitter side, not the drone side.

**Comment by stronnag** [comment]
INAV ready provides most of the infrastructure to do this:
* With `GCS NAV` flight mode modifier you can change the home position in flight
* With `GCS NAV` flight mode modifier you can implement follow me, including altitude control.

INAV (and EdgeTX) also provide / facilitate MSP over CRSF and S.port, so no infrastructure should be necessary, either on the FC or TX.

There are open source examples of doing all these things. Implementation requires doing some research and some LUA programming on the TX.

**Comment by dzaro-dev** [comment]
### Reasons for Required Changes:
Whilst it is appreciated that the infrastucture might be partially there, currently this is something that is only somewhat possible when connected to a Ground Control Station and there is no LUA script currently avalaible. MSP over CRSF and S.Port are part of the solution, but not the whole story. 

The reasons why some changes to the infrastrucutre are required are:

1. There would be additional flight modes, i.e. `RTO`, `FOLLOW-ME` (and variants of this).
2. There should be be two different home positions for the craft to know. One would be defined by the coordinates at take-off and this one would be fixed. The other would be defined as the coordinates of the operator, which would be changing and updated at some pre-defined interval. Hence distinguishing between `RTH` and `RTO`. 

### Additional Context:
There may be a case where the transmitter stops sending the required information for `RTO` or `FOLLOW-ME`, such as a long-term loss of GPS signal or transmitter signal, in which case the take-off location should still be available to the flight controller. Otherwise, if the home position is only moved via the MSP commands using `GCS NAV`, then in the event of a `FAILSAFE` due to a trasmitter malfunction, the craft will land at the last known / last updated `HOME` location, which may not be a safe place to attempt an automated landing. In such a situation, it would be better and safer for the craft to attempt to land where it took off. 

The configurator on the other hand would have to allow for failsafe behaviour to be a choise between `RTH` and `RTO` by the user. 

I hope this makes it a bit clearer in terms of how the feature should be implemented to cover all angles.

**Comment by MrD-RC** [comment]
I honestly don't think this is a good idea. DJI have all the collision detection sensors to try and stop their drones from hitting objects or people. INAV doesn't have that. So I would not want to see transmitter based follow me on safety grounds. If people really want it, Stronnag has a solution higher up.

**Reply by dzaro-dev** [reply]
So, just to get things right: you believe that attempting to land by going in a circular motion and then just heading out in a random direction is a better solution (i'm talking about fixed-wing aircraft)?

Here are just a few counter-arguments against your statement:

1. DJI collision detection can be turned off, it is just an optional default feature. `FOLLOW-ME` on DJI can work without collision detection.
2. Considering Remote ID requirements, it will have to be implemented sooner or later anyway. As you are in the UK, you are not concerned about this but there are many INAV users in the USA, who will need to have bi-directional exchange of GPS information. 
3. Unfortunately, stronag's solution is incomplete. As already mentioned above, it does not cover the initial feature proposal and it's implementation to the extent that it should be covered.

At the end of the day, `RTO` and `FOLLOW-ME` is something that many users will make great use of, if they had the feature. If you don't like to have this feature personally, you don't have to use it. You can ignore this developement and you don't have to build a GPS into your radio.  I honestly don't see a problem with a drone following the opearator at an adaptive altitude set by the operator. It is the same as DJI and no more or less safe than what DJI has (with or without collision detection). The point is that this feature would be available for both multirotors and fixed-wing aircraft.

**Reply by MrD-RC** [reply]
What you're talking about here has nothing to do with landing after RTH. That has nothing to to with follow me or using the GPS in a transmitter. Adding that would not change the land after RTH behaviour. 

I stick with my thoughts above.

**Reply by dzaro-dev** [reply]
Likewise, your statement about "DJI [having] all the collision detection sensors to try and stop their drones from hitting objects or people" has nothing to do with `RTO` and `FOLLOW-ME` flight modes and the proposed integration of bi-directional telemetry between transmitter and aircraft.

You may not like the idea of those two additional flight modes, (and again, that's just personal preference), but we are not here to debate which pizza toppings are better.

Also, as far as I understand INAV is not a personal project wholly owned by @MrD-RC or by myself, so the community will decide if they like this Feature Request or not.

**Reply by MrD-RC** [reply]
Hang on. You're the one here demanding something be added, that can already be done with a little research and implementation. Yet you're moaning at me for giving my opinion that it should not be implemented as a basic feature on safety grounds. 

No. I am not the owner or sole voice of INAV. But I am a contributor to the project, so have a right to express my opinion. 

You're also changing the goalposts to try and add validity to the request, that also don't fit. For RID and an enhanced landing approach, there is no need at all for a follow me mode. In fact, for landing, you would never want it to land exactly where the transmitter is. It would always need to be offset.

**Reply by b14ckyy** [reply]
I have to partially agree with darren here. 

@bishko first you talk about follow me and RTO. thats nice and in general I like that feature idea. But as Stronnag said, BOTH can be just realized by a LUA program on your radio. There is nothing that needs to be changed on INAV as INAV already provides all the necessary code. 

@stronnag has a side project with a Follow-Me "Wand" that acts as a GCS for INAV and can constantly change the loiter position relative to the device and therefore the operator to follow him. in INAV that's the GCS mode that is made for that. So you just need a LUA application in the radio that does the same as there already is a bidirectional communication between radio and INAV. 

Then you refer to autoland for fixed wings. Your feature request has NOTHING to do with autoland and would also not change the autoland feature at all. This is a whole different story and regarding that I am in contact with a developer already. of GPS is lost on the radio then the LUA program has to inform INAV to use the regular arming position or safehome as a home location again. Using the operators location as a landing position for wings is anyway a very stupid idea or do you want to stand in the middle of the field to land? 

Get someone who writes the LUA for you. No change in INAV needed for what you want. the software is ready for that, it just has to be used.

**Reply by dzaro-dev** [reply]
Sorry @b14ckyy, but the misunderstanding comes from the fact that @MrD-RC mentioned safety in light of `RTO` and `FOLLOW-ME`, where `RTO` would obviously include a landing procedure in the event of a failsafe or if defined by the user as the action to take after `RTH` or `RTO`. So, this may be a misunderstanding on my side, but then again, @MrD-RC just said one sentence about disagreeing without any substancial reasoning behind his concerns.

Fisrt of all, whilst the LUA script might work to lets say 50% of what it should, it does make things a lot more difficult for the average user to set up and if this feature is to be implemented at all, it should be as plug-and-play as possible, much like the auto-launch feature for example. 

To explain my idea a bit better, I would like to offer two example situations with 2 sub-scenarios each, where the suggestion by @stronnag would be limited as it is a workaourd and not a working solution:

### Example 1:
1a. Suppose you are riding a mountain bike down a hill and you would like the drone to follow you in FOLLOW-ME mode with adaptive altitude enabled. You take off, adjust the altitude of your drone to make sure that you always stay above the tree tops and lock it in place by enabling `FOLLOW-ME AD-ALT`. You then begin to cycle down the hill and about 3 miles down you stop and want to land (assuming it is open space and safe to do so). For whatever reason, you lose connection to your drone. Where does your drone go? Well, if you set up your failsafe procedure to be `RTH` that means going 3 miles back up the hill where it took off, right? Not convenient is it? 
1b. Imagine the same situation as above, but you lose connection to your drone whilst you were crossing some dense forrest area. In the event of a failsafe here with a moving RTH point as suggested by @stronnag , the drone would almost certainly crash because it will not be able to navigate and land through the dense forrest. 

### Example 2:
2a. Suppose you are in a car driving on an empty country road and would like the drone follow you in FOLLOW-ME mode with adaptive altitude enabled, just incase the road has some variation in altirude. As in the mountain bike example 1a, you take off, adjust your desired altitude offset and go into `FOLLOW-ME AD-ALT` mode. You get in your car and drive down the road for 5 miles but at the 5 mile mark you lose connection and the drone goes into failsafe mode. Using @stronnag 's approach with a changing home point, your drone attempts to land somewhere just behind you as that was the last GPS location that it had received before going into failsafe, but the location where it is landing is not safe for landing becuase there is a pond or a deep puddle of water or whatever. In this istuation the best option would be for the drone to go to the `SAFEHOME` location that you defined before your journey about 1 mile down the road. You continue your journey to that location, recover your drone, all is well.
2b. Lets say the drone has a total effective fligth range of 7 miles. As mentioned before, if you only have a moving `RTO` position defined as a moving `RTH` as suggested by @stronnag, you might land where you don't want to land in the event of a failsafe. On the other hand, if you have `RTH` as your failsafe procedure and the drone attemps to go back 5 miles, it will not make it because it will run out of battery on it's way to the take-off location. In this case you you will aslo crash because you planned your journery to be 6 miles, with 1 mile of range left for safety, but your drone will attempt to fly back 5 miles.  

This is the reason why we must distinguish between three different "return-to" positions, i.e. manually defined `SAFEHOME`, take-off defined `HOME` and transmitter defined `OPERATOR`, each defined and stored separately. Therefore, the moving RTH as a FOLLWO-ME substitute, and the automatic take-off defined `SAFEHOME` as a `RTH` subsiture wouldn't work very well in many situations. 

I do agree with you @b14ckyy , that having a LUA script to be able to test various scenarios so that one can come up with the safest and most optimal working sollution would be good as a starting point.

Maybe I am missing something here or there is a lack of understabnding from either side regarding the proposed `RTO` location, but I still belive that it would be better to have the the varieties of `HOME` locations, rather than the existing two doe to the reasons given above.

**Reply by b14ckyy** [reply]
To your Example 1: 
First of all: How the hell do you do a downhill bike ride with your TX16S Radio in your hand? 
Second: Downhill drone chase with no obstacle avoidance in a forest? 
Third: If the Drone is in follow me and within a few meters of you then why should it ever failsafe? If it failsafes then because of transmitter issues and in this case you want it to land in a safe spot and not hover over the ground in the middle of the forest until the battery dies. 
Sorry but this example is so full of flaws you can't even do that with a DJI drone as well. If you don't realize that you failsafed and it hovers there, you drive 2, 3 or 4 miles further downhill anfd then you realize your drone is gone, where do you want it to be? At the launch point or somewhere mid nowhere in a forest? 

And the same accounts for your second example. Leaving the drone hover mid nowhere without you realizing that it lost track of you is just a very bad idea. 

The only solution to that would be some kind of Rally Waypoints like ardupilot has, that you have to plan beforehand as a mission and offer the drone safe landing zones in case of a failsafe by going to the nearest one. But this is a completely different feature and has nothing to do with your initial Follow me feature request. 

without a rally waypoint setup the only thing that makes sense is: 

Failsafe goes to the arming home point
RTH goes to operator position if available, otherwise to arming home

Rally waypoints are only suitable if you plan the whole follow me trip beforehand as it is with ardupilot now.

**Reply by b14ckyy** [reply]
Oh and for my suggestion with the 2 RTH/RTO options, the only change in INAV would be the need for an additional RTO position that can be set and the decision based on RTH type (commanded or FS). That's it. The majority of the work has to be done on the radio to send the necessary data to the craft over MSP anyway. And before this is not done and working there is no reason to add this change to INAV IMHO. 

So get someone to write that LUA for you to provide the necessary data and THEN we can talk about feature changes like this in INAV. That's the smallest issue then.

**Reply by MrD-RC** [reply]
> the misunderstanding comes from the fact that @MrD-RC mentioned safety in light of `RTO` and `FOLLOW-ME`, where `RTO` would obviously include a landing procedure in the event of a failsafe or if defined by the user as the action to take after `RTH` or `RTO`. So, this may be a misunderstanding on my side, but then again, @MrD-RC just said one sentence about disagreeing without any substantial reasoning behind his concerns.
>

@bishko Are you kidding me. A drone flying in to someone using follow me because of no sensors or AI to try and avoid this happening, is not a substantial reason behind concerns! If you are personally happy to accept liability for people getting hurt because of this feature, then crack on.

> ### Example 1:
> 1a. Suppose you are riding a mountain bike down a hill and you would like the drone to follow you in FOLLOW-ME mode with adaptive altitude enabled. You take off, adjust the altitude of your drone to make sure that you always stay above the tree tops and lock it in place by enabling `FOLLOW-ME AD-ALT`. You then begin to cycle down the hill and about 3 miles down you stop and want to land (assuming it is open space and safe to do so). For whatever reason, you lose connection to your drone. Where does your drone go? Well, if you set up your failsafe procedure to be `RTH` that means going 3 miles back up the hill where it took off, right? Not convenient is it? 1b. Imagine the same situation as above, but you lose connection to your drone whilst you were crossing some dense forest area. In the event of a failsafe here with a moving RTH point as suggested by @stronnag , the drone would almost certainly crash because it will not be able to navigate and land through the dense forest.
> 

In this situation you'd be shit out of luck anyway. If it failsafes in this dense forest, and you can't regain control. It would try to land at the last point your mountain bike was at when it lost signal. Now lets say it does successfully make it through the trees, and not crash in to them. What's to say there won't be someone else on a mountain bike following you down the track. They then have a drone dropping down in front of them? What if there are hikers there. You have no control at the landing site, and the drone has no way of avoiding any potential people or objects. This is a big problem and a safety concern!

OK, lets reverse this scenario. You're out enjoying your ride in the forest. You're on a nice fast piece of singletrack, and having a great time. Then out of nowhere a 7" quad drops down in front of you (it has lost connection from someone's follow me setup). It's at chest height when you impact. How does your day look now?

> ### Example 2:
> 2a. Suppose you are in a car driving on an empty country road and would like the drone follow you in FOLLOW-ME mode with adaptive altitude enabled, just in case the road has some variation in altitude. As in the mountain bike example 1a, you take off, adjust your desired altitude offset and go into `FOLLOW-ME AD-ALT` mode. You get in your car and drive down the road for 5 miles but at the 5 mile mark you lose connection and the drone goes into failsafe mode. Using @stronnag 's approach with a changing home point, your drone attempts to land somewhere just behind you as that was the last GPS location that it had received before going into failsafe, but the location where it is landing is not safe for landing because there is a pond or a deep puddle of water or whatever. In this situation the best option would be for the drone to go to the `SAFEHOME` location that you defined before your journey about 1 mile down the road. You continue your journey to that location, recover your drone, all is well. 2b. Lets say the drone has a total effective flight range of 7 miles. As mentioned before, if you only have a moving `RTO` position defined as a moving `RTH` as suggested by @stronnag, you might land where you don't want to land in the event of a failsafe. On the other hand, if you have `RTH` as your failsafe procedure and the drone attempts to go back 5 miles, it will not make it because it will run out of battery on it's way to the take-off location. In this case you you will also crash because you planned your journey to be 6 miles, with 1 mile of range left for safety, but your drone will attempt to fly back 5 miles.
> 

Why would the last location not be "safe". If it is following you in a car, the last location would be on the road. Not in a pond. Obviously, it's not safe, as it is a road, so anyone can be using it. Stronnag's approach updating the home point means that if there is a valid safehome within range, it would be used. But don't forget, the whole point of a safehome is to be **safe**. Meaning you are there to oversee the landing. You can tell anyone there about the situation and make sure they are nowhere near the aircraft when it arrives. This is why they cannot be miles away. They are supposed to be close to home. Again, this is for safety.

**Comment by dzaro-dev** [comment]
### Responses:
@b14ckyy in response to your comments and questions:

> How the hell do you do a downhill bike ride with your TX16S Radio in your hand? 

Possible solutions:

1. You strap it to a backpack. 
2. You make a bike mount for a radio (front or rear)

Of course you would also make sure that you can lock all input, so that you don't do something unintentionally. 

Regarding the following:

> And the same accounts for your second example. Leaving the drone hover mid nowhere without you realizing that it lost track of you is just a very bad idea.

You do hear announcmenets from the radio such as: "Telemetry Lost" or "RSSI Critical", and with the Bluetooth mod, you could even hear it in your bluetooth earbuds. There are hunders of solutions to this problem. 


I agree with you on this one completely:

> Oh and for my suggestion with the 2 RTH/RTO options, the only change in INAV would be the need for an additional RTO position that can be set and the decision based on RTH type (commanded or FS). That's it. The majority of the work has to be done on the radio to send the necessary data to the craft over MSP anyway. And before this is not done and working there is no reason to add this change to INAV IMHO.

Essentially, this is all I was saying from the beginning. If we could distinguish between three different locations, it would make life a lot easier. In the mean time, I'll work on the MSP over CRSF, as suggested by @stronnag and maybe I'll manage to cobble something together. The only thing I would add is just a `FOLLOW-ME` mode in the modes tab so that if you have your RTH procedure set to "LAND" it doesn't actually land on you, but rather hovers at a preset altitude or at the altitude that you engage the `FOLLOW-ME`. The rest can be done in future versions. 

I menrtioned this previously here:

> This is the reason why we must distinguish between three different "return-to" positions, i.e. manually defined SAFEHOME, take-off defined HOME and transmitter defined OPERATOR, each defined and stored separately. Therefore, the moving RTH as a FOLLWO-ME substitute, and the automatic take-off defined SAFEHOME as a RTH subsiture wouldn't work very well in many situations.


@MrD-RC in response to your safety concerns: 

Safety is a function of many varibales, some of which we can control and other we can not. That's a fact of this hobby that none of us would be able to change. There are general safety concerns about any automated flight, not just `FOLLOW-ME` modes or `RTO` and this is irrespective of location or craft. 

Any power loss situations, such as failure of an ESC, a prop getting loose, etc. may result in a crash. `RTO` and `FOLLOW-ME` on their own do not have an impact on the chances of a catastropic failure and they certinaly do not increase the probability of causing a crash, compared to other automated flight modes.

Otherwise, that safety concern you have should be equal for all automated flight modes and in this case I would like to ask you: Are you indirectly proposing for all automated flight modes to be removed in future versions of INAV? - Kind of defeates the purpose of INAV, doesn't it.

If not, then the safety aspect of this debate is settled.

If yes, then please explain how `RTO` and `FOLLOW-ME` would be any more likely to cause personal injury or property damage compared to all other automated flight modes currently available in INAV and please provide evidence for your claims.


### Summary:
Adding `RTO` as a separate GPS coordinate for all future developements of this would be the first step towards a working solution. Hence, adding RTO as suggested by @b14ckyy would be a great startting point. The only thing to be added would be an accompanying `RTO` flight mode, in case the user wants to leave the `RTH` procedure working with the arm-defined location. Whatever is selected in the RTH, with landing or without, would only apply to `RTH` where as the `RTO` mode would have it's own follow-up procedure as defined by the user, i.e. with or without landing.

Safety concerns voiced by @MrD-RC and @b14ckyy over additional flight modes such as `RTO` and `FOLLOW-ME` are valid, but no have no more or less of an affect of overall safey compared to other existing automated flight modes. They are effectively dependent on careful planning and common sense, which can reduce or mitigate any residual risks by the pilot taking all reasonably necessary measures to avoid a crash in the first place.

**Reply by MrD-RC** [reply]
> ### Responses:
> @MrD-RC in response to your safety concerns:
> 
> Any power loss situations, such as failure of an ESC, a prop getting loose, etc. may result in a crash. `RTO` and `FOLLOW-ME` on their own do not have an impact on the chances of a catastrophic failure and they certainly do not increase the probability of causing a crash, compared to other automated flight modes.

@bishko You're completely missing the point. We're not talking about catastrophic failure here. No one said anything about dying ESCs or props getting loose. Of course they would have the same impact as anything else. Stop trying to muddy the water to defend `RTO` or `FOLLOW-ME` with irrelevant situations.

The situations that are important were raised, and you have completely ignored in your response. Of course `RTO` and `FOLLOW-ME` increase risk in failsafe situations (what we were talking about in your previous message. You even specifically used failsafes in your examples of why `RTO` and `FOLLOW-ME` would be good). We are going from a known home point or safehome, which is close to where we are, and we can assert some form of control over. To an unknown landing site, when we may not even realise the drone is going there. We have no control over who or what may be at that landing site. The chance of getting to the landing site to assert control over it is basically non-existent. First you have to notice that the drone is no longer following you. Then you have to stop, and try to find some way to getting a data or video link to the drone to find out what it's doing. The best you will get is a direction and distance. Then you have to turn around and try and get to it, using an approximation at best. At which point it could have already severely injured a mountain biker following you down that hill!

### Summary
To me, this is a gimmick feature to try and get some easy selfies. The safety aspect, which you seem to be ignoring, far out way the few benefits to very few pilots. If you fail to see the additional risk that an unknown home location, which you probably won't be at when the drone comes down. You really should forget all about this. Or if you do accept the risks, just get something from DJI.

**Reply by b14ckyy** [reply]
I make it short but this will be my last statement on this: 
Good luck finding any developer doing all these changes needed for INAV to make it work as you want it + keeping it somewhat safe (This request is a never-ending rabbit hole tbh as more and more needed changes will come up on the way) for maybe 4 or 5 people in the world ever using it this way.

You are better off with Ardupilot here as it already pretty much can do what you want. Or you just get a Skydio 2 Drone. That does EXACTLY what you want. Without a 900g Radio zip-tied to your mountain bike going down a forest path, jumping over stems and stones (just that picture in my head is ridiculous).

**Reply by dzaro-dev** [reply]
@MrD-RC And unfortunately you are missing my point as well. What's to stop someone from programming a mission that accidentaly flies a drone or plane into a building? Nothing. Anyone can do it, deliberately or by accident. These are the facts, and there is no muddying of any waters here. You will have to accept that at some point in time. Becuase you repeatedly fail to see that the risk probablility distribution is equal in all autonomos modes, there is no point to continue this conversation. `RTO` and `FOLLOW-ME` will happen. `RTO` would be the same as `RTH`, while`FOLLOW-ME` is just a dynamic version of that, meaning that they have the same level of risk, but you obviously fail to see that. 

PS: I have done a risk assessment of all automated flight modes and if one is to be objective, an equal risk can be found in all of them. I did not report my one-sided opinion (as you have), I reported the facts from both sides of the argument, acknowledging that there is a risk, but that it is not greater or smaller than the already existig risk imposed by other autonomous flight modes. 

@b14ckyy I don't know how you reached the conclusion of "4 or 5 people in the world ever", but I think that both you and I do not know how many people would be using a feature like this at this point in time, without having the feature in the first place. One thing is certain though: given the number of people using INAV, it's not going to be "4 or 5".
